Added detailed information for translators. (#1094)
jamulussoftware/jamulus: TRANSLATING.md

# Guide for Translators

This guide is written to provide instructions from scratch for contributing to the translation of the Jamulus application to other languages.

There are two parts:

1. Instructions for getting set up, which only needs to be done once.

2. The workflow for updating and contributing translations as a part of the preparation for a release of Jamulus.

----

> ## Setting up

Reviewer: This section feels more like a summary of the details that follow, rather than "Setting up", which implies a one-off activity.  I first though it was the "Part 1" from the "There are two parts".

Author: Good points, I’ll restructure slightly.

# Guide for Translators

This guide is written to provide instructions from scratch for contributing to the translation of the Jamulus application to other languages.

There are two parts:

1. Instructions for getting set up, which only needs to be done once.

2. The workflow for updating and contributing translations as a part of the preparation for a release of Jamulus.

----
## Setting up

The code for Jamulus is open source, and is managed and made available via the Github site.

> The translator must be able to do the following steps:

Reviewer 2: Given the pitch of the rest of this document, I wonder if this summary is perhaps counter-productive. If somebody who didn't know _how create an account on GitHub_ saw things like "Fork the Jamulus code" and "Clone their own repository" at the start, I wouldn't blame them for hitting the back button :-)

Author: Yes, fair point, I’ll have a think about it.

Instructions for use are in the [Qt Linguist Manual](https://doc.qt.io/qt-5/qtlinguist-index.html)

----
## Doing a translation

### Get the most up-to-date files and set up a new branch

#### With git command line tools

The first step is to get the local repo up to date with the upstream master:

```
cd projectdir/jamulus
git fetch upstream
```

This fetches information about the current state of the upstream repo. It is now necessary to apply any upstream changes to the local repo.
As the user will not be updating the `master` branch themselves, `git rebase` can be used to fast-forward to the current state:

```
git checkout master
git rebase upstream/master
git push
```

Finally, create a new branch for the changes that will be done (do not just do them on `master`):

```
git checkout -b v370-german
```

(The branch name is chosen by the user; the above name v370-german is an example for the German translation of V3.7.0)

The branch will be used later as the source of a Pull Request.

#### With Github Desktop

Select the current repository as `jamulus`, and the current branch as `master`.

Click on "Fetch origin" to get fully up to date with upstream.

> Drop down the "Current Branch" menu and click on the **New Branch** button. Give the branch a name such as `v370-german` and click **Create Branch**.

Reviewer 3: I'm wondering if we should use a consistent/different branch name. The official tags are r3_7_0. So maybe we should aim to match that style? Seperator-less version numbers will start to look strange if we ever have a 3.7.10 or 3.10.7.

```suggestion
Drop down the "Current Branch" menu and click on the **New Branch** button. Give the branch a name such as `translate-r3_7_0-german` and click **Create Branch**.
```

Note: There are other places below which use the branch name, so if it is changed, it should be changed there, too.

Do not click on **Publish branch** just yet.

### Work on the translation file

Open Qt Linguist, and navigate to the directory `src/res/translation` within your project directory.

In this directory are translation source files for each language, each with a `.ts` suffix. Don't worry about the `.qm` files, as they are compiled when building the release code.

Open the `.ts` file for the language being worked on.

> Each context in the left-hand column represents a source file or graphica form containing translatable strings. The icon beside each shows a green tick if there is nothing further to be done,

Reviewer 3: Not a native speaker, but is "string(s)" a word which a non-technical person would read as what we mean? Should we introduce this?

Also, graphica**l** lacks an l, I think.

```suggestion
Each context in the left-hand column represents a source file or graphical form containing translatable words and phrases ("strings"). The icon beside each shows a green tick if there is nothing further to be done,
```

a yellow tick if a translation needs attention, and a question mark if there are untranslated strings.

After clicking on a Context, the list of strings needing translation is shown, with a green tick beside the ones that have been done. The right-hand pane will show either the source code or the form,
so that the usage can be checked, in case this aids the translation.

Strings with a grey tick can be ignored. They are old strings that used to be used but are no longer. They are retained for reference in case they should be used again in the future.

Again, strings with a question mark require translation.

After some or all of the required strings have been translated, the file can be saved.

> ### Submit the updated translation as a PR

Reviewer 3: ```suggestion
### Submit the updated translation as a Pull Request (PR)
```

Once all translations have been done, the branch containing the changes must be commited to your own repo, pushed to Github (`origin`) and then a Pull Request raised to the upstream repo.

#### With git command line tools

First, check the current branch with `git status`, to make sure it is correct, and then do a commit:

```

> git commit -am 'Updated translations for v3.7.0'

Reviewer 3: ```suggestion
git commit -am 'Update German translations for v3.7.0'
```

Commit message should include the language and should be imperative-style (at least a quick scan of recent commit messages shows that this is the most common style).

```

Then do a `git push`.

It will probably tell you to set the upstream repository for tracking, and conveniently gives the command for copying and pasting:

```
git push --set-upstream origin v370-german
```

Finally, go to the Github website where it will most likely offer a banner saying there is a recent commit and offering to raise a Pull Request. Do so.

#### With Github Desktop

Select the current repository as `jamulus`, and the branch that was created above, such as `v370-german`.

The changed file(s) should be listed in the left-hand column as `src/res/translation/translation_xx_YY.ts`. When the file is selected, the differences will be displayed in the main panel.

> Add a simple commit message in the first box below the file list, (the "Update filename" box), and add any extra description in the Description box (optional, probably not required).

Reviewer 3: ```suggestion
Add a simple commit message in the first box below the file list, (e.g. change "Update filename" to something like "Update German translations for v3.7.0"), and add any extra description in the Description box (optional, probably not required).
```
